### ISSUE TITLE
[engine] Engine support for Consortium Party Primitives

### ISSUE THREAD

**Issue by nickchapman-da** [digital-asset/daml#15882]
## Support for consortium party primitives in Daml, LF and the engine.

Here is the [design document](https://docs.google.com/document/d/1Ha1u8hZF49U_DlQU_TLOm1K1uxOMGUgV2dTXT7lK41I/edit?pli=1#heading=h.ytuv3cfhm8o9).

See redesign document: _"Alternative Proposal: Baking authority changes into template choices"_ [here](https://docs.google.com/document/d/1Z8lbvTQcxik3_1rGpIie5Dw216M_grtwY2LA2Nn3zT8/edit#heading=h.4mr0cryy64m5)

Jump further down this issue for post-redesign comments & work [here](https://github.com/digital-asset/daml/issues/15882#issuecomment-1458062188)

### Overview

A consortium is represented by a Party in the Daml language. The essence of a consortium is that it acts with the combined authority of its members, which are also represented as Parties in Daml. A consortium party will have special rules for validation: say 2/3 of its member parties will be required to validate. Also, the membership of a consortium can change dynamically over time. But neither of these aspects are directly visible or represented at the Daml language level. More importantly, a consortium party is unable to submit to the ledger, because unlike other parties, a consortium party does not reside on any given participant. So without any new functionality it would be impossible to obtain an active-contract signed by a consortium party.

The proposed new functionality is for a Daml-LF builtin which dynamically obtains the authorization of a consortium party when executed in the context of the authorization of the consortium members. From the point of view of the Daml Engine, obtaining the authority of a consortium party requires delegating the question to the enclosing ledger, and recording the answer in the constructed transaction.

### Daml

The new functionality will be exposed to the user via a new Daml primitive:
```
withAuthorityOf : [Party] -> Update a -> Update a
```

The behaviour of `withAuthorityOf parties` will attempt to place  the set of authorizing parties to the given `parties`, based on the authority of the current authorization context . For more precise behavior, see [summary](https://github.com/digital-asset/daml/issues/15882#issuecomment-1352914931) below.

### Example

Here is a small example, slightly simplified from the one given in the design document, so that the `Accept` and `Ratify` stages are made explicit. It assumes a new primitive `withAuthorityOf` with the above type.

```
template HasConsortiumAutority
  with
    consortiumParty: Party
  where
    signatory consortiumParty

template ProposeConsortiumAuthority
  with
    proposer: Party
    accepted: [Party]
    obs: [Party]
    consortiumParty: Party
  where
    signatory proposer, accepted
    observer obs

    choice Accept : ContractId ProposeConsortiumAuthority
      with who: Party
      controller who
      do create this with accepted = who :: this.accepted

    choice Ratify : ContractId HasConsortiumAutority controller proposer
      do
        withAuthorityOf accepted $ do
          withAuthorityOf [consortiumParty] $ do
            create HasConsortiumAutority with consortiumParty

```

### Implementation checklist

- [x] New Daml primitive: `withAuthorityOf`
- [x] (Haskell) compilation in `LFConversion/Primitives.hs`
- [x] (Haskell) type checker
- [x] (Haskel) simplifier
- [x] (Haskell) `.dar` encode(decode)
- [x] extension of LF `.proto` format
- [x] (Scala) `.dar` decode(encode)
- [x] New LF primitive `BWithAuthorityOf` in `Ast.scala`
- [x] Internal LF parser `ExprParser.scala`
- [x] Scala type checker `Typing.scala`
- [x] Scala utils: `AstRewriter`, `ExprIterable` etc
- [x] Compilation to speedy `PhaseOne.scala`
- [x] New speedy builtin: `SBWithAuthorityOf` in `SBuiltin.scala`
- [x] Implementation of `SBWithAuthorityOf`
- [x] New speedy result/callback in `SResult.scala`
- [x] New engine result/callback in `engine/Result.scala`
- [x] Support in `PartialTransaction.scala` and `Transaction.scala`
- [x] TX `.proto` changes
- [x] Revisit possible impact of serialization size for choice with default authorization (Double check  with @andreaslochbihler-da) 
- [x] fix bug in the calculation of the current authorization context in `SBuiltin`
- [ ] ~improve `daml-script` driven testing~
- [x] support `choiceAuthorizers` in TX encode/decode! (and test)
- [x] updating TX Value generator
- [ ] #16810
- [ ] updating LF/TX `.rst` spec files
- [ ] ~Question: Do we need a Haskell side feature flag?~

**Comment by andreaslochbihler-da** [comment]
> `actingAsConsortium : [Party] -> Party -> Update ()`

Whether a set of parties can act in the name of a consortium should be upward-closed, i.e., is a set of parties `X` suffices, then any bigger set should also suffice. So conceptually, the Daml code need not specify the parties whose authority is being used to obtain the consortium's authority, i.e., the function

```
actAsParty : Party -> Update ()
```

would also do. Daml has a different take on making the required authorizers explicit:

* Choices compute the controller explicitly from the input contract instance and the choice argument. So in a sense, the required authorizers of an exercise are explicit.
* `fetch` does not specify who actually authorizes the contract retrieval. The actors on a Fetch node are computed dynamically from the authorization context.

@meiersi-da @moritzkiefer-da Since you're going to be the first users of this primitive, can you provide some input on whether the relevant consortium members should be passed explicitly?

**Comment by cocreature** [comment]
I'd be inclined to not pass the relevant parties explicitly. If we want something like that, a generic `restrictAuthority` primitive seems like the better choice than special casing this here.

I think `actingAsConsortium : [Party] -> Party -> Update ()` is the wrong type though. We should imho scope this so something like:
```
withConsortiumAuthority : Party -> Update a -> Update a
```

**Comment by andreaslochbihler-da** [comment]
Regarding the naming: consortium parties are indistinguishable from normal Daml parties inside the language. Yet, the name `actingAsConsortium` conveys the impression that party given as a second argument must be a consortium party, but this depends on how the ledger answers Daml engine's questions. So for future-proofing, I suggest to not use the name `consortium` in any of the user-facing Daml functions. How about `actingAsParty` or `withAuthority` instead?

Then, when we come up with other use cases for one party granting their authority to someone else, we can simply reuse this language primitive and merely extend the ledger with the rules accordingly.

**Comment by andreaslochbihler-da** [comment]
> We may want to consider limiting the scope of the extended authority, perhaps by offering the following primitive, either as well as, or instead of actingAsConsortium

I agree with @moritzkiefer-da that a scoped primitive feels more intuitive to me because authority is currently passed always in scoped versions.

**Comment by andreaslochbihler-da** [comment]
> recording the answer in the constructed transaction.

Daml engine can and should cache these answers during an interpretation: If the same primitive call happens several times during an interpretation, the ledger is asked only once.

This means that there's an advantage to specifying the members whose authority is used, rather than deriving that from the authorization context. Since the authorization context may differ in different call sites and yet the same subset of members would suffice, then we will record multiple queries and their answers.

Unfortunately, a ledger cannot compute a unique minimal set from an authorization context for a consortium party (there may be several quorums that meet the 2/3 threshold). It could compute a set of sets of parties that would work, but there may be quadratically many such sets for a consortium, so this doesn't sound like a good approach either.

@moritzkiefer-da This is a small indication that explicitly specifying the parties may have some benefit. Not sure whether that's enough to swing you the other way?

**Comment by andreaslochbihler-da** [comment]
> recording the answer in the constructed transaction.

Please also record the node IDs of the exercise nodes whose choice bodies performed the call to `actingAsConsortium`. This is important so that we know which projections in Canton should include a recorded answer and where we should not ship it around (for subtransaction privacy).

**Comment by nickchapman-da** [comment]
Another argument in favour of explicitly specifying the parties is to allow writing examples such as above, where the `proposer` is not forced to implicitly `Accept`.

To gain the convenience of the implicit version, perhaps we could have a Daml primitive to discover the current parties whose authority is in scope.

**Comment by meiersi-da** [comment]
> ```
> withConsortiumAuthority : Party -> Update a -> Update a
> ```

Looks too weak: wouldn't that mean that we can only execute `Update`s that are authorized as that single party?

What about
```
withAuthorityOf : [Party] -> Update a -> Update
```
which would check that the surrounding authorizers grant authority for acting as these parties *and* at the same time restrict the authorizers for the inner update. 

> Daml engine can and should cache these answers during an interpretation: If the same primitive call happens several times during an interpretation, the ledger is asked only once.

@andreaslochbihler-da : what's the rationale for caching this in Daml engine instead of having the ledger introduce an in-memory cache if needed? Note that I don't expect our code to need that, as we have a single call to `withAuthorityOf` as part of bootstrapping the consortium.

**Comment by cocreature** [comment]
> Looks too weak: wouldn't that mean that we can only execute Updates that are authorized as that single party?

I was assuming the consortium party gets added but the other parties don’t get removed. But I can see how that may be confusing.

`withAuthorityOf` sounds nice and also works for non-consortium parties like `restrictAuthority` would.

**Comment by nickchapman-da** [comment]
Thinking about the transaction tree needed to record `withAuthority`, we will require a new TX node, because authority is fixed for each given node.

This is specially true for the scoped version of the operator, but even for the unscoped version, we would require something like a new node to indicate where the increased authority begins.

@andreaslochbihler-da or am I missing something?

**Comment by cocreature** [comment]
> because authority is fixed for each given node.

why is that a requirement? We need the authorizers of a node to start from but I don’t see why we cannot change the authority within a node.

**Comment by andreaslochbihler-da** [comment]
@meiersi-da 's proposal of `withAuthorityOf` that restricts the authority to the given set of parties is IMO the best proposal so far. The semantics should be:
* If the given parties `P` are not a subset of the current authorization context `C`, then Daml engine asks the ledger whether `C` suffices to authorize the parties in `P - C`. If the ledger responds no, then interpretation fails with an uncatchable exception.
* Otherwise, or if the given parties `P` are a subset of the current authorization context, then the authorization context is set to `P` for the execution of the given `Update`.
* Afterwards, the authorization context is restored to `C`.

@nickchapman-da Does this create significantly more work than the proposal you wrote?

> what's the rationale for caching this in Daml engine instead of having the ledger introduce an in-memory cache if needed?

It's not strictly necessary. It just felt in line with Daml's contract cache; Daml engine queries the ledger during a single interpretation for each contract only once, even if there are multiple fetch nodes for the same contract, but this is probably motivated by caching the validation that Daml engine needs to do upon a fetch. It's fine to ignore this.



> Thinking about the transaction tree needed to record `withAuthority`, we will require a new TX node, because authority is fixed for each given node.
> 
> This is specially true for the scoped version of the operator, but even for the unscoped version, we would require something like a new node to indicate where the increased authority begins.

I don't think it is necessary to create a node for these things. I'm happy to discuss this in a call.

**Comment by nickchapman-da** [comment]
> @nickchapman-da Does this create significantly more work than the proposal you wrote?
I don't think it has much bearing on how much work is involved.

> I don't think it is necessary to create a node for these things. I'm happy to discuss this in a call.
Yes, I think that would be helpful.

**Comment by nickchapman-da** [comment]
Thanks you everyone for your input. To summarise...

The new operator is not specific to consortium parties, and so should be named more generally. The consensus seems to be for `withAuthorityOf`.

Choices for: `withAuthorityOf`
- A: extent of effect: (scoped vs unscoped) -- SCOPED
- B: on whose authority the request is made (implicit vs explicit) -- IMPLICIT
- C: authority granted (replace vs extend) -- REPLACE

```
withAuthorityOf :: [Party] -> Action a -> Action a
```

with the semantics as described by Andreas above:

- If the given parties `P` are not a subset of the current authorization context `C`, then Daml engine asks the ledger whether C suffices to authorize the parties in `P - C`. If the ledger responds no, then interpretation fails with an uncatchable exception.
- Otherwise, or if the given parties `P` are a subset of the current authorization context, then the authorization context is set to `P` for the execution of the given Update.
- Afterwards, the authorization context is restored to `C`.

Because the semantics are to replace and not extend the current authority context, the operator can also be used to restrict the current authority.

For validation, we wont need a new node type in the transaction tree to capture exactly where calls to `withAuthorityOf` are made. Instead we can just record (on the side; per exercise-node) the set of _authority-requests_ made to the ledger during the construction of this node.

An _authority-request_ of the ledger is of the form: `C` suffices for `P-C`.

Note specifically that in the case that authority is being restricted (`P` is a subset of `C`) then no request will be made to the ledger, and nothing will be recorded in the transaction.

**Comment by nickchapman-da** [comment]
Issue description and example updated.

**Comment by andreaslochbihler-da** [comment]
@nickchapman-da @remyhaemmerle-da Having thought a bit more about this, I'm not sure whether we're making the right trade-off on whether `withAuthorityOf` should become a node in the tree. Let me revisit that here.

The discussion so far focused on minimizing the implementation effort for `withAuthorityOf` for Daml language.  We haven't really considered the implementation effort for pulling this through the Canton synchronization protocol, but I suspect that adding new `withAuthority` nodes to the tree vs. recording the queries on the side will not make a big difference for Canton. Recording the queries on the side means that we cannot check well-authorization of a transaction tree without running Daml engine any more:

- The recorded queries do not show where `withAuthorityOf` is used to restrict the authorization context. So this information is not visible in the transaction tree or its metadata, but it's vital to check well-authorization.
- The recorded information currently merely mentions the parent node where the query was needed; it doesn't specify which child nodes are within the extended authorization scope, but the authorization check should take this into account.

The argument basically is that one can hide this stuff inside the model-conformance check. But this means that well-authorization and model conformance check become entangled. The discussion is still going on whether there is much practical value in talking about well-authorized transactions of which one does not know whether they are model-conformant. On the theory side, there is definitely value, though:

`withAuthorityOf` actually requires an extension of the ledger model, and this is currently missing from the task list above.

If we don't add corresponding nodes to the trees, this will be very ugly to do, because the ledger model explicitly distinguishes between authorization and model-conformance - and we are making a big fuzz about this.

Therefore, I suggest to **strongly** reconsider the alternative of introducing separate `WithAuthority` nodes:

```
case class WithAuthority(child: ImmArray[NodeId]) extends Node
```

Like `Rollback` nodes, `WithAuthority` nodes are not action nodes and therefore do not have informees. The projection rules for `WithAuthorityOf` are simple: `WithAuthority` nodes disappear above the root level nodes of the projection:
```
project p (WithAuthority children) = mapConcat (project p) children
```
Clearly, `WithAuthority` nodes remain inside actions that are not included in the projection.


Note that `withAuthorityOf` distributes over `bind` on the `Update` monad. That is,
`withAuthorityOf ps $ do { x <- upd; f x }` is semantically equivalent to
```
do
  x <- withAuthorityOf ps upd
  withAuthorityOf ps (f x)
```
This means that, `WithAuthority` does **not** need to be scoped across a whole subtree, unlike `Rollback` nodes and we could normalize `WithAuthority` nodes so that each of them gets only a single child node.
Unfortunately, this is not as simple though because `withAuthorityOf` does not vanish on a `pure`: `withAuthorityOf ps (pure x)` is **not** equivalent to `pure x`. This means we cannot drop childless `WithAuthorityOf` nodes like we do for rollback normalization.

So an alternative definition would be
```
case class WithAuthority(child: Option[NodeId]) extends Node
```
i.e., making the child optional.

Yet, `WithAuthority` also interacts with normalization of `Rollback` nodes:
```
Rollback (WithAuthority ps None) -> WithAuthority ps None
```
because a failed authority exception cannot be caught. It is not yet clear to me to what extent we want to normalize `Rollback` and `WithAuthority`. What I am fairly certain about, however, is that it is not a good idea to simply introduce a `WithAuthority` node whenever interpretation comes across a `withAuthorityOf` primitive, because this would bloat our trees with unnecessary information. For example,
```
withAuthorityOf ps (withAuthorityOf ps ...)
```
should give the same tree as
```
withAuthorityOf ps ...
```

This list of open design questions shows that we should spend more work on getting this concept of authorization context change into the ledger model. To unblock the implementation work, I think we can (for now) stick with the original proposal (record the queries in the transaction metadata in LF language version dev) so that we can evaluate whether this is the right feature. In parallel, we should rethink the impact on the ledger model as our theoretical foundation.

**Comment by nickchapman-da** [comment]
This issue needs updating following the redesign: _"Alternative Proposal: Baking authority changes into template choices"_ [here](https://docs.google.com/document/d/1Z8lbvTQcxik3_1rGpIie5Dw216M_grtwY2LA2Nn3zT8/edit#heading=h.4mr0cryy64m5)

**Comment by nickchapman-da** [comment]
Provisional Implementation Checklist:
- [ ] New driving example in Daml
- [ ] New `authorizers` (or `acting`) field in TX `Node.Exercise`
- [ ] Default value for `authorizers`: `signatories` + `controllers`
- [ ] Compilation pipeline: Daml syntax -> LF -> engine -> speedy -> TX
- [ ] Update tests at speedy/lf and daml-script level
- [ ] Kill `Node.Authority` with fire

**Comment by nickchapman-da** [comment]
### Driving example for _choice authority_
```
template HasConsortiumAutority
  with
    consortiumParty: Party
  where
    signatory consortiumParty

template ProposeConsortiumAuthority
  with
    proposer: Party
    accepted: [Party]
    obs: [Party]
    consortiumParty: Party
  where
    signatory proposer, accepted
    observer obs

    choice Accept : ContractId ProposeConsortiumAuthority
      with who: Party
      controller who
      do create this with accepted = who :: this.accepted

    choice Ratify : ContractId HasConsortiumAutority
      controller proposer
      do exercise self Ratify1

    choice Ratify1 : ContractId HasConsortiumAutority
      authority accepted -- restrict authority
      controller proposer
      do exercise self Ratify2

    choice Ratify2 : ContractId HasConsortiumAutority
      authority consortiumParty -- gain authority
      controller proposer
      do create HasConsortiumAutority with consortiumParty
```

**Comment by meiersi-da** [comment]
> ### Driving example for _choice authority_
> (Replacing the `withAuthorityOf` example given in the issue description.)
> 
> ```
> template HasConsortiumAutority
>   with
>     consortiumParty: Party
>   where
>     signatory consortiumParty
> 
> template ProposeConsortiumAuthority
>   with
>     proposer: Party
>     accepted: [Party]
>     obs: [Party]
>     consortiumParty: Party
>   where
>     signatory proposer, accepted
>     observer obs
> 
>     choice Accept : ContractId ProposeConsortiumAuthority
>       with who: Party
>       controller who
>       do create this with accepted = who :: this.accepted
> 
>     choice Ratify : ContractId HasConsortiumAutority
>       controller proposer
>       do exercise self Ratify1
> 
>     choice Ratify1 : ContractId HasConsortiumAutority
>       authority accepted -- restrict authority
>       controller proposer
>       do exercise self Ratify2
> 
>     choice Ratify2 : ContractId HasConsortiumAutority
>       authority consortiumParty -- gain authority
>       controller proposer
>       do create HasConsortiumAutority with consortiumParty
> ```

Two questions:
1. Can the `authority` clause refer to choice argument parameters?
2. Have you considered writing the `authority` clause as the last step before `do` and this way avoid the potential confusion that `controller`'s must be a subset of the parties in the `authority` clause?

**Comment by nickchapman-da** [comment]
> Can the authority clause refer to choice argument parameters?
Yes. This will be allowed.

**Comment by nickchapman-da** [comment]
> Have you considered writing the authority clause as the last step before do ...

The syntax will be flexible, allowing `controller`, `observer` and `authority` declarations in any order.
The `controller` will be mandatory, the others optional. Repeats will be errors.

**Comment by nickchapman-da** [comment]
One more point about the syntax...

Prior to this work on _choice authority_, there was no `where` keyword required or allowed to introduce the optional `observer` decl and the mandatory `controller` decl. This had the rather unsatisfactory consequence that the `observer` decl (if present) *must* precede the `controller` decl. With the addition of an optional `authority` decl, it is very hard to find a satisfactory grammar without the introduction of the `where` clause.

I was very happy to see that Andreas's example had already made use of the `where` keyword (perhaps without even realising that this was not the current syntax). Introducing the `where` keyword makes the syntax much cleaner and more consistent with other parts of the Daml grammar which have declaration lists (for example template declarations).

My proposal is that we introduce the `where` keyword as the canonical syntax for declarations scoped to a `choice`. (And update user facing docs accordingly). But we continue to support the legacy syntax (no `where` keyword) so all previously allowed syntax remains legal. i.e. `controller` only choices, and `observer`+`controller` choices.

Use of `where` is mandatory when making an `authority` declaration. Use of `where` is mandatory to permit free order of decls, i.e. an `observer` declaration to follow a `controller` declaration.

**Comment by nickchapman-da** [comment]
@meiersi-da 

I'm not sure I understand the semantic restriction assumed by your 2nd question:

> Have you considered writing the authority clause as the last step before do and this way avoid the potential confusion that controller's must be a subset of the parties in the authority clause?

Why must controllers be a subset of authority parties?

**Comment by nickchapman-da** [comment]
Actually, looking more closely at my example... there is a mistake in the `controller` for `choice Ratify2`. I had written `proposer`, but on reflection it should be `accepted`, to match the restricted authority in scope at the `exercise` call made by `choice Ratify1`.

(example updated)

**Comment by nickchapman-da** [comment]
Updated example so `Ratify` and `Ratify1` choices are `nonconsuming`.

**Comment by nickchapman-da** [comment]
Actually there is a further problem with this example.

The example attempts to show how `alice` (the `proposer`) can restrict her authority via an explicit `authority` declaration in choice `Ratify1`, so that her authority is not in scope in `Ratify2`. But the problem is that her authority is in scope anyway because she is a signatory on the template that contains `Ratify2`.

We should be able to fix this by separating out this logic into a separate template.

**Comment by leonidr-da** [comment]
1. In the original example, why do we need a `withAuthorityOf accepted` in the `Ratify` choice, don't we have their authority already as `accepted` are signatories? Is this to restrict the set of authorizing parties?
2. I would discourage the new choice style with `authority` keyword as it creates "private" choices; they're part of the public/external API but should only be called by other choices.

**Comment by andreaslochbihler-da** [comment]
> 1. In the original example, why do we need a `withAuthorityOf accepted` in the `Ratify` choice, don't we have their authority already as `accepted` are signatories? Is this to restrict the set of authorizing parties?

Yes. The point is to restrict the authorizing parties for the inner `withAuthorityOf`.

> 2. I would discourage the new choice style with authority keyword as it creates "private" choices; they're part of the public/external API but should only be called by other choices.

Why do you consider them "private"? A choice with `authority` clause changes the authorization context in the body of the choice. It doesn't change anything about who's allowed to exercise the choice (the controller).

**Comment by leonidr-da** [comment]
> Yes. The point is to restrict the authorizing parties for the inner withAuthorityOf.

**Comment by nickchapman-da** [comment]
> Yes. The point is to restrict the authorizing parties for the inner withAuthorityOf.

Just to be super clear. There is no `withAuthorityOf` function. That was the earlier design 1 which we abandoned.
The design2 which we implemented has an optional `authority` declaration on any choice.

> Awesome! Does this mean we would be able to control who is able to see what's inside 

I don't think so. The `authority` decl has no affect on the visibilty, but only an affect on the authority which is in scope,

**Comment by dylant-da** [comment]
I think Nick finished everything he wanted with this - @andreaslochbihler-da where did explicit authority end up w.r.t Nick's engine changes here?

**Comment by paulbrauner-da** [comment]
@andreaslochbihler-da feel free to re-open.
